### PR TITLE
feat: messages between parcels

### PR DESCRIPTION
**Pull request by menduz**
- Separates comms into Broker and World implementations
- Adds a WS broker
- Fixes chat analytic events
- Avoids removal of internal persistent scenes in CLI preview
- Separates hellmap to avoid continously builds
- adds a `make watch-single FILE="-100.80*"` to only build a subset of scenes
- adds a `make initialize-ecs-npm-link` to prepare and link a local `decentraland-ecs` to test with the CLI in real time
- uses a new version of decentraland-rpc that fixes a bug in the lifecycle hooks of remote controllers
- adds the original `.proto` files to the comms project
- separate the single `.proto` into `Broker` and `World` (same as our services)

- [x] Update .proto files and regenerate
- [x] Separate into communication channels (WS/Comms)
- [x] Tests

### CODE REVIEW

File: packages/decentraland-ecs/types/dcl/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,5817 @@
+
```
Comment (eordano) [inline packages/decentraland-ecs/types/dcl/index.d.ts]: @menduz should this file be commited? It looks auto-generated to me

File: packages/shared/comms/BrokerConnection.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+        if (msgType === MessageType.WEBRTC_ICE_CANDIDATE) {
+          try {
+            await this.webRtcConn!.addIceCandidate({ candidate: sdp })
+          } catch (err) {
+            this.logger.error(err)
+          }
+        } else if (msgType === MessageType.WEBRTC_OFFER) {
+          try {
+            this.logger.log('1')
```
Comment (hugoArregui) [inline packages/shared/comms/BrokerConnection.ts]: remove this? (maybe I left it here)

**Review by hugoArregui — APPROVED** [review]

File: Makefile [hunk 3]
```diff
@@ -27,7 +29,7 @@ compile-dev:
 build-sdk: build-support
 	@echo "$(GREEN)======================= Building SDK =======================$(RESET)"
 	$(COMPILER) build.sdk.json
-	cd $(PWD)/packages/decentraland-ecs; $(PWD)/node_modules/.bin/api-extractor run --typescript-compiler-folder "$(PWD)/node_modules/typescript" --local
```
Comment (eordano) [inline Makefile:30]: consider dropping `--verbose`?

File: Makefile [hunk 4]
```diff
@@ -37,9 +39,14 @@ build-support:
...
 	$(COMPILER) build.support.json
 
+build-hell-map:
+	@echo "$(GREEN)===================== Building hell map ====================$(RESET)"
+	$(COMPILER) build.hell-map.json
+
 build-test-scenes: build-sdk
 	@echo "$(GREEN)=================== Building test scenes ===================$(RESET)"
 	$(COMPILER) build.test-scenes.json
+	$(MAKE) build-hell-map
```
Comment (eordano) [inline Makefile:49]: Why are we making a distinction on this set of scenes? I think the granularity is kind of broken/I'd use a different name than "hell map" (`perf-test-scenes`?)

File: Makefile [hunk 5]
```diff
@@ -117,10 +124,29 @@ lint-fix:
 	node_modules/.bin/prettier --write 'packages/**/*.{ts,tsx}'
 	node_modules/.bin/prettier --write 'packages/decentraland-ecs/types/dcl/index.d.ts'
 
+watch: export NODE_ENV=development
```
Comment (eordano) [inline Makefile:127]: oops?

File: packages/config/index.ts [hunk 6]
```diff
@@ -85,6 +85,7 @@ export const EDITOR: boolean = !!(global as any).isEditor
 // Development
 export const AVOID_WEB3: boolean = !!(global as any).avoidWeb3 || EDITOR
 export const DEBUG = location.search.indexOf('DEBUG') !== -1 || !!(global as any).mocha || PREVIEW || EDITOR
+export const DEBUG_LOCAL_COMMS = location.search.indexOf('LOCAL_COMMS') !== -1 || PREVIEW
```
Comment (eordano) [inline packages/config/index.ts]: How do you feel about `s/DEBUG_LOCAL_COMMS/USE_LOCAL_COMMS`?

File: packages/decentraland-ecs/api-extractor.json [hunk 7]
```diff
@@ -1,25 +1,325 @@
+/**
```
Comment (eordano) [inline packages/decentraland-ecs/api-extractor.json:1]: Verbose as F*. 

*No generated code shall be commited to repos*

File: packages/decentraland-ecs/src/decentraland/Types.ts [hunk 8]
```diff
@@ -1,12 +1,15 @@
 import { ReadOnlyVector3, ReadOnlyQuaternion } from './math'
 
+/** @public */
```
Comment (eordano) [inline packages/decentraland-ecs/src/decentraland/Types.ts:3]: @menduz what's up with all these jsdocs?

File: packages/decentraland-ecs/types/@decentraland/CommunicationsController/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+declare module '@decentraland/CommunicationsController' {
```
Comment (eordano) [inline packages/decentraland-ecs/types/@decentraland/CommunicationsController/index.d.ts:1]: should this file be .gitignored?

File: packages/decentraland-ecs/types/env/es5.d.ts [hunk 10]
```diff
@@ -19,8 +19,8 @@ and limitations under the License.
 /// ECMAScript APIs
```
Comment (eordano) [inline packages/decentraland-ecs/types/env/es5.d.ts:19]: Should this file be `.gitignored`?

File: packages/decentraland-ecs/types/tsconfig.json [hunk 11]
```diff
@@ -20,7 +20,6 @@
       "@decentraland/ChatController",
       "@decentraland/EthereumController",
       "@decentraland/SocialController",
-      "@decentraland/EntityController",
```
Comment (eordano) [inline packages/decentraland-ecs/types/tsconfig.json:23]: Oops?

File: packages/shared/apis/CommunicationsController.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+      message,
+      sender: sender.uuid
+    })
+  }
+
+  @exposeMethod
+  async send(message: string): Promise<void> {
+    sendParcelSceneCommsMessage(this.cid, message)
+  }
+}
```
Comment (eordano) [inline packages/shared/apis/CommunicationsController.ts:44]: Beautifully short and effective

**Review by eordano — COMMENTED** [review]
All nits, I still haven't reviewed the meaty part.

File: packages/shared/comms/BrokerConnection.ts [hunk 13]
```diff
@@ -0,0 +1,306 @@
...
+  constructor(public url: string) {
+    this.connectRTC()
+    this.connectWS()
+    // TODO: reconnect logic, handle disconnections
+  }
+
+  logDebug(): void {
+    if (this.ws && this.ws.readyState === SocketReadyState.OPEN) {
+      const state = (this.authenticated ? 'authenticated' : 'not authenticated') + `my alias is ${this.alias}`
+      this.logger.log(state)
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts:54]: .debug?

File: packages/shared/comms/BrokerConnection.ts [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+  public reliableDataChannel: RTCDataChannel | null = null
+  public unreliableDataChannel: RTCDataChannel | null = null
+
+  public stats: Stats | null = null
+
+  public logger: ILogger = createLogger('Broker: ')
+
+  public onMessageObservable = new Observable<BrokerMessage>()
+
+  get hasUnreliable() {
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts]: `s/hasUnreliableChannel`

File: packages/shared/comms/BrokerConnection.ts [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+    return (this.unreliableDataChannel && this.unreliableDataChannel.readyState === 'open') || false
+  }
+
+  get hasReliable() {
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts]: `s/hasReliableChannel`

File: packages/shared/comms/BrokerConnection.ts [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+      this.webRtcConn.onicecandidate = null
+      this.webRtcConn.ondatachannel = null
+      this.webRtcConn.close()
+      this.webRtcConn = null
+    }
+
+    if (this.ws) {
+      this.ws.onmessage = null
+      this.ws.onerror = null
+      this.ws.onclose = null
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts:87]: :/ really? shouldn't we call it first?

File: packages/shared/comms/BrokerConnection.ts [hunk 17]
```diff
@@ -0,0 +1,306 @@
...
+          try {
+            await this.webRtcConn!.setRemoteDescription(new RTCSessionDescription({ type: 'answer', sdp }))
+          } catch (err) {
+            this.logger.error(err)
+          }
+        }
+        break
+      }
+      default: {
+        if (this.stats) {
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts:192]: This should be outside of the switch

File: packages/shared/comms/BrokerConnection.ts [hunk 18]
```diff
@@ -0,0 +1,306 @@
...
+    this.ws.send(bytes)
+  }
+
+  private connectRTC() {
+    this.webRtcConn = new RTCPeerConnection({
+      iceServers: commConfigurations.iceServers
+    })
+
+    this.webRtcConn.onsignalingstatechange = (e: Event) => {
+      this.logger.log(`signaling state: ${this.webRtcConn!.signalingState}`)
```
Comment (eordano) [inline packages/shared/comms/BrokerConnection.ts:221]: .debug?

File: packages/shared/comms/CliBrokerConnection.ts [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+  private ws: WebSocket | null = null
+
+  constructor(public url: string) {
+    this.connectWS()
+  }
+
+  logDebug(): void {
+    if (this.ws && this.ws.readyState === SocketReadyState.OPEN) {
+      const state = (this.alias ? 'authenticated' : 'not authenticated') + `my alias is ${this.alias}`
+      this.logger.log(state)
```
Comment (eordano) [inline packages/shared/comms/CliBrokerConnection.ts:34]: .debug?

File: packages/shared/comms/CliBrokerConnection.ts [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+      throw new Error('trying to message using null unreliable channel')
+    }
+    this.sendCoordinatorMessage(data)
+  }
+
+  close() {
+    if (this.ws) {
+      this.ws.onmessage = null
+      this.ws.onerror = null
+      this.ws.onclose = null
```
Comment (eordano) [inline packages/shared/comms/CliBrokerConnection.ts:58]: Warning: we should call this first

File: Makefile [hunk 21]
```diff
@@ -130,8 +156,27 @@ only-watch:
 			"node ./scripts/buildECSprojects.js --watch" \
 			"node ./scripts/test.js --keep-open"
 
+# initializes a local dev environment to test the CLI with a linked version of decentraland-ecs
+initialize-ecs-npm-link: build-sdk
```
Comment (fmiras) [inline Makefile:159]: This is awesome

File: packages/shared/comms/debug.ts [hunk 22]
```diff
@@ -136,22 +136,18 @@ export class Stats {
...
-      const url = connection.url
-      if (connection.ws && connection.ws.readyState === SocketReadyState.OPEN) {
-        const state =
-          (connection.authenticated ? 'authenticated' : 'not authenticated') + ` - my alias is ${connection.alias}`
-        if (connection.ping >= 0) {
-          log(`  ${url}, ping: ${connection.ping} ms (${state})`)
-        } else {
-          log(`  ${url}, no ping info available (${state})`)
-        }
+      connection.connection.logDebug()
```
Comment (eordano) [inline packages/shared/comms/debug.ts]: What does this do?

File: packages/shared/comms/IBrokerConnection.ts [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+  channel: string
+}
+
+export interface IBrokerConnection {
+  onMessageObservable: Observable<BrokerMessage>
+  readonly hasUnreliable: boolean
+  readonly hasReliable: boolean
+  sendReliable(data: Uint8Array): void
+  sendUnreliable(data: Uint8Array): void
+  logDebug(): void
```
Comment (eordano) [inline packages/shared/comms/IBrokerConnection.ts]: Is `logStateInfo()` a more descriptive name?

File: packages/config/index.ts [hunk 24]
```diff
@@ -139,7 +140,10 @@ export function getServerConfigurations() {
     landApi: `https://api.decentraland.${TLDDefault}/v1`,
     content: `https://content.decentraland.${TLDDefault}`,
     worldInstanceUrl: `wss://world-comm.decentraland.${TLDDefault}/connect?method=noop`,
-    darApi: 'https://schema-api-staging.now.sh/dar'
+    darApi:
+      TLDDefault === 'zone' || TLDDefault === 'today'
+        ? 'https://schema-api-v2.now.sh/dar'
```
Comment (fmiras) [inline packages/config/index.ts:145]: Can we track with either an issue or a TODO task to setup .zone environment for this?

File: packages/shared/comms/index.ts [hunk 25]
```diff
@@ -195,26 +250,31 @@ export function onPositionUpdate(context: Context, p: Position) {
...
-
-        topic = `profile:${hash}`
-        rawTopics += topic + ' '
-
-        topic = `chat:${hash}`
-        rawTopics += topic
-
-        if (x !== xMax || z !== zMax) {
-          rawTopics += ' '
+        if (!rawTopics.includes(hash)) {
```
Comment (eordano) [inline packages/shared/comms/index.ts:256]: Thank you <3

File: public/test-parcels/-100.100.gamekit-cube/game.ts [hunk 26]
```diff
@@ -26,6 +27,10 @@ export class RotatorSystem implements ISystem {
   }
 }
 
+const bus = new MessageBus()
+
+log('hello a a s')
```
Comment (eordano) [inline public/test-parcels/-100.100.gamekit-cube/game.ts:32]: 🙁

**Review by eordano — CHANGES_REQUESTED** [review]

File: packages/shared/comms/index.ts [hunk 27]
```diff
@@ -195,26 +250,31 @@ export function onPositionUpdate(context: Context, p: Position) {
...
 
-    worldConnection.updateSubscriptions(rawTopics)
+    currentParcelTopics = rawTopics.join(' ')
+  }
+
+  const parcelSceneSubscriptions = getParcelSceneSubscriptions()
+
+  const parcelSceneCommsTopics = parcelSceneSubscriptions.join(' ')
+
+  let topics = currentParcelTopics + (parcelSceneCommsTopics.length ? ' ' + parcelSceneCommsTopics : '')
```
Comment (fmiras) [inline packages/shared/comms/index.ts]: could this be `const` instead?

Comment (fmiras) [inline packages/shared/comms/index.ts]:
```suggestion
  const topics = currentParcelTopics + (parcelSceneCommsTopics.length ? ' ' + parcelSceneCommsTopics : '')
```

File: Makefile (hunk 3 above)
Comment (menduz) [reply Makefile:30]: It is on purpose, it behaves like an --info tho. It really helped while debugging, so I kept it

File: Makefile (hunk 4 above)
Comment (menduz) [reply Makefile:49]: Can we tackle that renaming issue in a different PR?

File: Makefile (hunk 5 above)
Comment (menduz) [reply Makefile:127]: On purpose, it avoids minification during watch's first stage (`compile-dev`)

File: packages/config/index.ts (hunk 24 above)
Comment (menduz) [reply packages/config/index.ts:145]: https://github.com/decentraland/open-sea-schema-api/issues/6

File: packages/decentraland-ecs/api-extractor.json (hunk 7 above)
Comment (menduz) [reply packages/decentraland-ecs/api-extractor.json:1]: I found that very useful while setting it up. I kept it there for future generations, should we remove it?

File: packages/decentraland-ecs/src/decentraland/Types.ts (hunk 8 above)
Comment (menduz) [reply packages/decentraland-ecs/src/decentraland/Types.ts:3]: It marks the annotated declarations to be exported in the .d.ts rollup made by api-extractor. Therefore, every `@public` thing is available in the ECS global context

File: packages/decentraland-ecs/types/@decentraland/CommunicationsController/index.d.ts (hunk 9 above)
Comment (menduz) [reply packages/decentraland-ecs/types/@decentraland/CommunicationsController/index.d.ts:1]: Nope, it is hand written, it specifies the interface of the modules being loaded by the RPC against the client

File: packages/decentraland-ecs/types/tsconfig.json (hunk 11 above)
Comment (menduz) [reply packages/decentraland-ecs/types/tsconfig.json:23]: That file was removed

File: packages/shared/comms/BrokerConnection.ts (hunk 16 above)
Comment (menduz) [reply packages/shared/comms/BrokerConnection.ts:87]: In most scenarios yes, we have a TODO to handle reconnections with closes and errors. That may lead us to endless loops. Anyways, we can rearrange it now and change it later. We only have to remember it

File: packages/shared/comms/BrokerConnection.ts (hunk 17 above)
Comment (menduz) [reply packages/shared/comms/BrokerConnection.ts:192]: It changes depending the case
```
this.stats.others.incrementRecv
this.stats.webRtcSession.incrementRecv
```

File: packages/shared/comms/BrokerConnection.ts (hunk 18 above)
Comment (menduz) [inline packages/shared/comms/BrokerConnection.ts:221]:
```suggestion
      this.logger.debug(`signaling state: ${this.webRtcConn!.signalingState}`)
```

File: packages/config/index.ts (hunk 6 above)
Comment (menduz) [reply packages/config/index.ts]: done

**Review by fmiras — APPROVED** [review]

File: public/test-parcels/-100.100.gamekit-cube/game.ts (hunk 26 above)
Comment (menduz) [reply public/test-parcels/-100.100.gamekit-cube/game.ts:32]: Believe it or not, we have a test tied to that `log` 💫

File: packages/decentraland-ecs/types/dcl/index.d.ts (hunk 1 above)
Comment (menduz) [reply packages/decentraland-ecs/types/dcl/index.d.ts]: File removed.

File: packages/decentraland-ecs/types/env/es5.d.ts (hunk 10 above)
Comment (menduz) [reply packages/decentraland-ecs/types/env/es5.d.ts:19]: Nope, it is the file used to provide the environment ECS+ES5 for "import-less" scenes

File: packages/shared/comms/debug.ts (hunk 22 above)
Comment (menduz) [reply packages/shared/comms/debug.ts]: Addressed

File: packages/shared/comms/IBrokerConnection.ts (hunk 23 above)
Comment (menduz) [reply packages/shared/comms/IBrokerConnection.ts]: `printDebugInformation` it is

File: packages/shared/comms/BrokerConnection.ts (hunk 14 above)
Comment (menduz) [reply packages/shared/comms/BrokerConnection.ts]: addressed

File: packages/shared/comms/BrokerConnection.ts (hunk 15 above)
Comment (menduz) [reply packages/shared/comms/BrokerConnection.ts]: addressed

File: packages/shared/comms/index.ts (hunk 27 above)
Comment (menduz) [reply packages/shared/comms/index.ts]: yes

File: packages/shared/comms/CliBrokerConnection.ts (hunk 19 above)
Comment (menduz) [reply packages/shared/comms/CliBrokerConnection.ts:34]: there is no .debug in our logger yet :/

**Comment by eordano** [comment]
Overall:
- The hellmap moves could be their own PR
- The api-extractor update would be great as a different PR

File: Makefile (hunk 4 above)
Comment (BrianAmadori) [reply Makefile:49]: If we move this to another PR, maybe we should remove the hell map scenes while we are at it. I found the builder scenes in zone -64,64 are way better for benchmarking.

**Review by eordano — APPROVED** [review]

File: packages/decentraland-ecs/api-extractor.json (hunk 7 above)
Comment (eordano) [reply packages/decentraland-ecs/api-extractor.json:1]: It's generated code. No generated code shall be commited. It's noise.

File: packages/decentraland-ecs/types/env/es5.d.ts (hunk 10 above)
Comment (eordano) [reply packages/decentraland-ecs/types/env/es5.d.ts:19]: Did you write it all? If it's generated code, it should be removed, if it comes from somewhere else we could make it a dependency

File: packages/shared/comms/BrokerConnection.ts (hunk 16 above)
Comment (eordano) [reply packages/shared/comms/BrokerConnection.ts:87]: Please add a TODO

**Pull request merged**